### PR TITLE
Add sale order lot selection

### PR DESCRIPTION
**Pull request by eLBati**
Following https://github.com/OCA/sale-workflow/pull/140

### CODE REVIEW

File: sale_order_lot_selection/__openerp__.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+# You should have received a copy of the                                #
+# GNU Affero General Public License                                     #
+# along with this program.  If not, see <http://www.gnu.org/licenses/>. #
+#                                                                       #
+#########################################################################
+
+
+{
+    'name': 'Sale Order Lot Selection',
+    'version': '1.0',
```
Comment (ajite) [inline sale_order_lot_selection/__openerp__.py]: Please check the guidelines on how to define the module version number.
https://github.com/OCA/maintainer-tools/blob/master/CONTRIBUTING.md#version-numbers

**Comment by ajite** [comment]
LGTM technically (once the module version number is correct).

Now it seems obvious but I never thought of assigning a variable to self.env[model] like you did on line 82 of sale_order_lot_selection/model/sale.py. I would have assigned it to `None` but your way is much better. I will use it :).

**Comment by eLBati** [comment]
@ajite thanks, fixed

File: sale_order_lot_selection/view/sale_view.xml [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+        <record id="view_order_tree_lot" model="ir.ui.view">
+            <field name="name">sale.order.tree.lot</field>
+            <field name="model">sale.order</field>
+            <field name="inherit_id" ref="sale.view_order_form"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='order_line']/tree/field[@name='product_id']" position="after">
+                    <field name="lot_id" domain="[('product_id','=', product_id)]" context="{'product_id': product_id}"/>
+                </xpath>
+                <xpath expr="//field[@name='order_line']/form/group/group/field[@name='product_id']" position="after">
+                    <field name="lot_id" domain="[('product_id','=', product_id)]" context="{'product_id': product_id}"/>
```
Comment (gurneyalex) [inline sale_order_lot_selection/view/sale_view.xml]: the field `lot_id` should only be visible to the members of the `stock.group_production_lot` res.groups

**Comment by gurneyalex** [comment]
when testing on runbot, I noticed a weird interaction with sale_rental: when you edit a sale order line with both modules installed there is a crash:

Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/service/model.py", line 113, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/addons/web/controllers/main.py", line 944, in call_kw
    return self._call_kw(model, method, args, kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/api.py", line 241, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/api.py", line 363, in old_api
    result = method(recs, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/models.py", line 5880, in onchange
    record._onchange_eval(name, field_onchange[name], result)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/models.py", line 5798, in _onchange_eval
    method_res = getattr(self._model, method)(_args)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/api.py", line 241, in wrapper
    return old_api(self, *args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/api.py", line 363, in old_api
    result = method(recs, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/addons/sale_rental/rental.py", line 219, in product_id_change_with_wh_with_rental
    flag=flag, warehouse_id=warehouse_id)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3114033-202-d04e4f/openerp/api.py", line 239, in wrapper
    return new_api(self, _args, *_kwargs)
TypeError: 'NoneType' object is not callable

This looks like a weird interaction with the redefinition of the on change in both modules, which use a different decorator. Maybe @alexis-via could share an insight here.

**Comment by gurneyalex** [comment]
other remark: I can make a SO with a product qty which is greater than the available qty for the specified prodlot, and this does not raise any warning. Maybe it's intentional but IMO this deserves a mention in the README.

**Comment by alexis-via** [comment]
@gurneyalex I tried on my laptop with both sale_rental and sale_order_lot_selection, but I was not able to reproduce the crash... I didn't try on runbot yet.

**Comment by rvalyi** [comment]
@alexis-via just in case, sometimes I saw inconsistencies between runbot and local tests because if I remember well on Runbot the module are first installed with the -i option and only then the tests are run with options -u and --test-enable so some data from the previous installation may make some strange side effects. Eventually that could be the reason why it would be different on Runbot, worth testing.

**Comment by eLBati** [comment]
@gurneyalex thanks for review.
See last 3 commits.
In particular, about https://github.com/OCA/sale-workflow/pull/202#issuecomment-137560058, I wrote a test that covers the described use case, but the module correctly raises exception (see https://github.com/eLBati/sale-workflow/blob/add_sale_order_lot_selection/sale_order_lot_selection/tests/test_sale_order_lot_selection.py#L259-L262).
So maybe did you mean something different?

**Comment by alexis-via** [comment]
I was able to reproduce the crash yesterday ; I fixed it (I used @api.v7 on product_id_change in sale_rental instead of @api.multi) and I also developped a "compatibility" module to make sale_rental compatible with sale_order_lot_selection.

Please give me a few days to clean this up and sumbit it.

**Comment by eLBati** [comment]
@alexis-via the `product_id_change_with_wh` compatibility should be fixed with https://github.com/eLBati/sale-workflow/commit/438a4a6352b5909d3e83dff3287ecabd1b3e3b85

**Comment by alexis-via** [comment]
@eLBati Good ! I didn't see this recent commit ! I was working in the train yesterday without Internet connection and my branch was not up-to-date. This commit should avoid the cash, but it's not enough to make the 2 modules compatible from a "functionnal" point of view. So I'll update my code to put back the @api.multi on product_id_change and propose a module to make "sale_rental" compatible with your module (it also involves introducing some hooks in your module)

**Comment by alexis-via** [comment]
hopefully, the "sale" module uses the new API in v9, so these kind of problems will be gone !

**Comment by alexis-via** [comment]
As promised, I made a PR to eLBati's branch to make sale_rental and sale_order_lot_selection really compatible (not just from a technical point of view, but also from a functional point of view !). The PR is here https://github.com/eLBati/sale-workflow/pull/4
And there is a screencast here:
http://people.via.ecp.fr/~alexis/screencast-rental-lot-selection.ogv

**Comment by gurneyalex** [comment]
@eLBati ok the warning does trigger. I probably had my product set as Consumable when I did my first test. 

:+1:

File: sale_order_lot_selection/view/sale_view.xml [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+<openerp>
+    <data>
+
+        <record id="view_order_tree_lot" model="ir.ui.view">
+            <field name="name">sale.order.tree.lot</field>
+            <field name="model">sale.order</field>
+            <field name="inherit_id" ref="sale.view_order_form"/>
+            <field name="arch" type="xml">
+                <xpath expr="//field[@name='order_line']/tree/field[@name='product_id']" position="after">
+                    <field name="lot_id" domain="[('product_id','=', product_id)]" context="{'product_id': product_id}" groups="stock.group_production_lot"/>
```
Comment (StefanRijnhart) [inline sale_order_lot_selection/view/sale_view.xml]: Why set 'product_id'? Do you mean 'default_product_id' or is that equivalent nowadays?

**Comment by rafaelbn** [comment]
@gurneyalex do this module still need fixing? Please change label if not.
@eLBati I will review this module when tecnhical changes and reviews are done. Thanks

File: sale_order_lot_selection/view/sale_view.xml (hunk 3 above)
Comment (eLBati) [inline sale_order_lot_selection/view/sale_view.xml]: It seems equivalent: they both work. Nevertheless, I changed to `default_product_id` as it is clearer

**Comment by StefanRijnhart** [comment]
Thanks, looks really good! :+1:

**Comment by eLBati** [comment]
@rafaelbn still some work in progress to make this module work together with `sale_rental`: https://github.com/eLBati/sale-workflow/pull/4

I don't know in details what @alexis-via means saying the 2 module are functionally incompatible, because i don't know `sale_rental` enough.

However, while waiting for https://github.com/eLBati/sale-workflow/pull/4, I cited the incompatibility in README, to not block this PR

**Comment by alexis-via** [comment]
@eLBati You can merge my PR https://github.com/eLBati/sale-workflow/pull/4 and it will become compatible with sale_rental. I invested a lot of work on this PR, so please merge it and we can move forward. Please, don't merge this PR before https://github.com/eLBati/sale-workflow/pull/4 is merged, to respect my work.

**Comment by eLBati** [comment]
PR https://github.com/eLBati/sale-workflow/pull/4 needs fixing: https://github.com/eLBati/sale-workflow/pull/4#issuecomment-152156545

**Comment by alexis-via** [comment]
@eLBati I wanted to work on it tonight, but I lost my evening fixing stupid bugs like this one https://github.com/OCA/bank-statement-import/pull/41
I hope my next evening will be more quiet and I will have time to work on my sale_rental PR.

**Comment by eLBati** [comment]
Hi @alexis-via , do you plan to work on https://github.com/eLBati/sale-workflow/pull/4?
Thanks

**Comment by eLBati** [comment]
I would merge this and let `sale_rental` compatibility to be fixed with another PR

**Comment by alexis-via** [comment]
yes, go ahead, don't wait for me.

**Pull request merged**